main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1081 @@
+        c.expectTypeM(expected = Type.mkVector(tvar, loc), actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = Type.Int32, actual = tpe2, exp2.loc)
+        c.unifyTypeM(evar, Type.mkUnion(eff1, eff2, loc), loc)
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.VectorLength(exp, loc) =>
+        val elmVar = Type.freshVar(Kind.Star, loc)
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(Type.mkVector(elmVar, loc), tpe, exp.loc)
+        val resTpe = Type.Int32
+        val resEff = eff
+        (resTpe, resEff)
+
+      case Expr.Ref(exp1, exp2, tvar, evar, loc) =>
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val regionType = Type.mkRegion(regionVar, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        c.expectTypeM(tpe2, regionType, exp2.loc)
+        c.unifyTypeM(tvar, Type.mkRef(tpe1, regionVar, loc), loc)
+        c.unifyTypeM(evar, Type.mkUnion(eff1, eff2, regionVar, loc), loc)
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.Deref(exp, tvar, evar, loc) =>
+        val elmVar = Type.freshVar(Kind.Star, loc)
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val refType = Type.mkRef(elmVar, regionVar, loc)
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(expected = refType, actual = tpe, exp.loc)
+        c.unifyTypeM(tvar, elmVar, loc)
+        c.unifyTypeM(evar, Type.mkUnion(eff, regionVar, loc), loc)
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.Assign(exp1, exp2, evar, loc) =>
+        val elmVar = Type.freshVar(Kind.Star, loc)
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val refType = Type.mkRef(elmVar, regionVar, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        c.expectTypeM(expected = refType, actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = elmVar, actual = tpe2, exp2.loc)
+        c.unifyTypeM(evar, Type.mkUnion(eff1, eff2, regionVar, loc), loc)
+        val resTpe = Type.Unit
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.Ascribe(exp, expectedTpe, expectedEff, tvar, loc) =>
+        // An ascribe expression is sound; the type system checks that the declared type matches the inferred type.
+        val (actualTpe, actualEff) = visitExp(exp)
+        c.expectTypeM(expected = expectedTpe.getOrElse(Type.freshVar(Kind.Star, loc)), actual = actualTpe, loc)
+        c.unifyTypeM(actualTpe, tvar, loc)
+        c.expectTypeM(expected = expectedEff.getOrElse(Type.freshVar(Kind.Eff, loc)), actual = actualEff, loc)
+        val resTpe = tvar
+        val resEff = actualEff
+        (resTpe, resEff)
+
+      case Expr.InstanceOf(exp, _, _) =>
+        val (_, eff) = visitExp(exp)
+        c.expectTypeM(expected = Type.Pure, actual = eff, exp.loc)
```

Comment:
I don't think this should be there, why would it matter that the expression being checked is pure? if it is something impure you could just letbind it so I'm not sure what it achieves. @magnus-madsen ?

Reply:
Agree. It should not require purity. Good catch!

Reply:
listed